[WPE][GTK] Introduce WPE 2.0 API version, and ENABLE_2022_GLIB_API flag to indicate either GTK 4 or WPE 2.0

mcatanzaro opened the pull request:
#### b2a6414f90b356961962b1e57ad3a83405d2a4a2
<pre>
[WPE][GTK] Introduce WPE 2.0 API version, and ENABLE_2022_GLIB_API flag to indicate either GTK 4 or WPE 2.0
<a href="https://bugs.webkit.org/show_bug.cgi?id=244641">https://bugs.webkit.org/show_bug.cgi?id=244641</a>

Reviewed by Adrian Perez de Castro.

Since there are few users of the WPE 1.1 API, and we don&apos;t need to keep
it around, WPE 2.0 will simply replace WPE 1.1. In contrast, the WPE 1.0
API is still required until Igalia is ready to phase out libsoup 2.

With a new API version, I will be able to start breaking things in
subsequent commits.

I&apos;ve also fixed some misleading code in OptionsWPE.cmake where option
dependencies were located under an unrelated comment, intermixed with
option definitions.

* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/254653@main">https://commits.webkit.org/254653@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/898bec6aa0b1e7a2ad84f6f7978f90be5855769d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34315 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99099 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32809 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93449 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95411 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76602 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/82123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/80948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81425 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30560 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76286 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30309 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26436 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3266 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33760 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/38776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/78871 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32471 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17282 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -238,8 +239,8 @@ elseif (WEBKITGTK_API_VERSION VERSION_EQUAL "4.1")
-    CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(WEBKIT 0 0 0)
-    CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(JAVASCRIPTCORE 0 0 0)
+    CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(WEBKIT 1 0 1)
+    CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(JAVASCRIPTCORE 1 0 1)
```
Why did you bump library version here?

mcatanzaro replied:
It's in case this PR is ready to land before https://github.com/WebKit/WebKit/pull/3834 is ready. That PR is still blocked by https://github.com/WebKit/WebKit/pull/3858 (which is awaiting your review, when you have time for it), whereas this PR is ready with no other dependencies.

Basically the goal is that once this PR lands, I can start making API changes.

mcatanzaro replied:
BTW when #3834 lands, I will reset this back to `0 0 0` since it would be a new API name.

Also, I considered submitting a separate pull request to get rid of CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE, since using libtool triples just seems rather confusing to me. But since you and Adrian make the releases, it's really only your opinions that matter for that. Curious what you think.

mcatanzaro replied:
Don't need to change the library version anymore because https://github.com/WebKit/WebKit/pull/3834 will land first, and that will change the API name.

carlosgcampos commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -157,9 +157,10 @@ if (USE_SOUP2)
     set(SOUP_API_VERSION 2.4)
     set(WPE_API_VERSION 1.0)
 else ()
+    SET_AND_EXPOSE_TO_BUILD(ENABLE_2022_GLIB_API ON)
```
I guess this is to have a common ifdef for new api common to gtk and wpe. Adding an unused variable in this patch looks weird to me in any cae, I would add it in the first patch actually using it.

mcatanzaro replied:
Yes, we discussed this in https://bugs.webkit.org/show_bug.cgi?id=242409 and several times in our private chat. At least, sort of. The discussion felt rather one-way. :P

I wanted to do this in a separate PR so that we can agree on and land the new build flag distinct from any particular API changes.

carlosgcampos commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -157,9 +157,10 @@ if (USE_SOUP2)
     set(SOUP_MINIMUM_VERSION 3.0.0)
     set(SOUP_API_VERSION 3.0)
-    set(WPE_API_VERSION 1.1)
+    set(WPE_API_VERSION 2.0)
```
Is the assumption that nobody is using wpewebkit with soup3 accurate? I think it would be safer to simply add a new api version like for GTK.

mcatanzaro replied:
That's fine, but then I need to expose ENABLE_2022_GLIB_API as a public build option for WPE. @aperezdc, is this OK?

After previous discussions with @aperezdc, I gathered that he was primarily concerned about maintaining API stability for libsoup 2 API for a short while longer. My understanding is that WPE is OK with API breaks much more frequently than GTK is, and therefore we'll soon be able to remove everything except the new WPE API version.

mcatanzaro replied:
BTW this was previously decided in https://bugs.webkit.org/show_bug.cgi?id=242407, but it seems harmless to change our minds and keep it for a little while.

mcatanzaro replied:
Need a final decision on this, please.

mcatanzaro replied:
For now we'll keep both 1.0 and 1.1 APIs. I think they should both be removed soon as it should be easy to port to 2.0, but Igalia can do that when desired.

I will make 2.0 the default API and add a new option ENABLE_WPE_1_1_API to get the old version.

carlosgcampos commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -175,7 +176,7 @@ endif ()
     CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(WEBKIT 21 0 18)
 else ()
-    CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(WEBKIT 2 0 2)
+    CALCULATE_LIBRARY_VERSIONS_FROM_LIBTOOL_TRIPLE(WEBKIT 3 0 3)
```
New api version is incompatible change so we should do C + 1, 0, 0. But still, I think it should be safer to dd a new api version instead of changing existing and released 1.1 one.

mcatanzaro replied:
Ah, you're right. I assumed I knew how to use the libtool versions without looking it up... shame on me.

Anyway, since this patch changes the name of the API, the generated library name will be totally different, so I can simply reset it down to `0 0 0`, OK?

mcatanzaro replied:
> Anyway, since this patch changes the name of the API, the generated library name will be totally different, so I can simply reset it down to 0 0 0, OK?

Will do this.

carlosgcampos reviewed (commented):
I'm not sure I understand the library version bumps, so I think this patch should just add the new api version for wpe. We can add the new api ifdef in a follow up when actually needed and used.

mcatanzaro committed:
[WPE][GTK] Introduce WPE 2.0 API version, and ENABLE_2022_GLIB_API flag to indicate either GTK 4 or WPE 2.0
https://bugs.webkit.org/show_bug.cgi?id=244641

Reviewed by Adrian Perez de Castro.

Since there are few users of the WPE 1.1 API, and we don't need to keep
it around, WPE 2.0 will simply replace WPE 1.1. In contrast, the WPE 1.0
API is still required until Igalia is ready to phase out libsoup 2.

With a new API version, I will be able to start breaking things in
subsequent commits.

I've also fixed some misleading code in OptionsWPE.cmake where option
dependencies were located under an unrelated comment, intermixed with
option definitions.

* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/254653@main
